helix-term/src/ui/editor.rs

```diff
@@ -380,6 +385,84 @@ impl EditorView {
         spans
     }
 
+    pub fn render_context(
+        editor: &Editor,
+        doc: &Document,
+        view: &View,
+        surface: &mut Surface,
+        theme: &Theme,
+    ) {
+        if let Some(syntax) = doc.syntax() {
+            let text = doc.text().slice(..);
+            let tree = syntax.tree();
+            let byte = doc.selection(view.id).primary().cursor(text);
+
+            let mut parent = tree
+                .root_node()
+                .descendant_for_byte_range(byte, byte)
+                .and_then(|n| n.parent());
+
+            // context is list of numbers of lines that should be rendered in the LSP context
+            let mut context: Vec<usize> = Vec::new();
+
+            while let Some(curr) = parent {
+                let line = text.byte_to_line(curr.start_byte());
+
+                // if parent of previous node is still on the same line, use the parent node
+                if let Some(&prev_line) = context.last() {
+                    if prev_line == line {
+                        context.pop();
+                    }
+                }
+
+                context.push(line);
+                parent = curr.parent();
+            }
```

Comment:
We should also apply some filtering based on queries. Taking inspiration from https://github.com/nvim-treesitter/nvim-treesitter-context: `impl_item`, `struct`, `enum`, `method`, `function` for Rust (`class`, `function`, `'method` being good defaults I guess).
This should be configurable in `grammar` sections of the languages.toml file I think.

Reply:
Added option to configure `sticky-context-nodes` per language and added example config for rust. Once we are sure about the naming/approach I will try to add config for more languages.

Reply:
Sweet! Thank you!